### PR TITLE
Use the dask distributed scheduler

### PR DESCRIPTION
**Pull request by sjperkins**
- [ ] Tests added / passed

  ```bash
  $ py.test --flake8 -v -s .
  ```

  If the flake8 tests fail, the quickest way to correct
  this is to run `autopep8` and then `flake8`
  to fix the remaining issues.

  ```
  $ pip install -U autopep8 flake8
  $ autopep8 -r -i .
  $ flake8 .
  ```

### CODE REVIEW

**Comment by sjperkins** [comment]
@paoloserra This branch forces full garbage collection after each chunk of visibilities has been written. Can you try it and see whether it improves the memory leak issue?

**Comment by sjperkins** [comment]
Run this a bit on my laptop didn't seem to help.

**Comment by sjperkins** [comment]
Might need to consider jemalloc

**Comment by sjperkins** [comment]
@paoloserra I can run a simulation through with testcb.ms on my laptop on the dask distributed scheduler.

```bash
$ $ crystalball -sm ~/data/edb_2-sources.txt ~/data/testcb.ms/ -mf 0.25
```

Note the low memory fraction because I'm already using half of my laptop's memory.

By contrast, the standard dask ThreadPool scheduler does seem to Out Of Memory on the above problem.

My guess for why this happens is that the ThreadPool scheduler runs through the graph in a constrained order, whereas the distributed scheduler is more [sophisticated](https://docs.dask.org/en/latest/scheduling.html) about how it executes the problem, especially in regards to freeing up memory.

Can you give this a test on your box?

**Comment by paoloserra** [comment]
The test on testcb.ms with edb_2-sources.txt goes well but it's not very challenging. It completes in ~40 min on my desktop.

```
-------------------------------------------
system RAM = 62.78 GB
nr of logical CPUs = 8
nr sources = 4603
nr rows    = 349530
nr chans   = 98
nr corrs   = 2
sources per chunk = 115 (auto setting)
rows per chunk    = 11590 (auto setting)
expected memory usage = 31.14 GB
-------------------------------------------
```

A more challening test with 

```
-------------------------------------------
system RAM = 62.78 GB
nr of logical CPUs = 8
nr sources = 32380
nr rows    = 501921
nr chans   = 478
nr corrs   = 2
sources per chunk = 52 (auto setting)
rows per chunk    = 5248 (auto setting)
expected memory usage = 31.10 GB
-------------------------------------------
```

is currently executing and started swapping after some ~20 min. Swap is slowly but steadily increasing, and I expect it will eventually crash. I will keep it running to see whether that happens indeed, and report here.

Note that, compared to the first test, the main change in the second test is in the nr of sources (7x), while the nr of rows didn't increase a lot (1.4x). This is in line with previous runs, where I've noticed that working on the same MS but with a lot more sources caused memory issues (with the same crystalball settings).

**Comment by sjperkins** [comment]
@paoloserra That's unfortunate.

> This is in line with previous runs, where I've noticed that working on the same MS but with a lot more sources caused memory issues (with the same crystalball settings).

Yes, it's probably the number of source chunks that is causing this:

In the first case: 4603 // 115 ~ 40
In the second case: 32380 // 52 ~ 623

There are large parallel reductions happening over the source axis. I guess that the traversal of the graph by the scheduler is not optimal in terms of memory. I'll have to see if I can make this smarter in some way.

**Comment by paoloserra** [comment]
> A more challening test with
> 
> ```
> -------------------------------------------
> system RAM = 62.78 GB
> nr of logical CPUs = 8
> nr sources = 32380
> nr rows    = 501921
> nr chans   = 478
> nr corrs   = 2
> sources per chunk = 52 (auto setting)
> rows per chunk    = 5248 (auto setting)
> expected memory usage = 31.10 GB
> -------------------------------------------
> ```
> 
> is currently executing and started swapping after some ~20 min. Swap is slowly but steadily increasing, and I expect it will eventually crash.

... which it did. See error message below.
```

[###                                     ] | 9% Completed |  1hr 49min 49.5sdistributed.worker - WARNING -  Compute Failed
Function:  execute_task
args:      ((subgraph_callable, array([37, 37, 37, ..., 44, 44, 44]), None, (array([[15737,   125],
       [15863,   306],
       [16170,   243],
       [16414,   197],
       [16612,   311],
       [16924,   209],
       [17134,   350],
       [17485,  3507]], dtype=int32), array([   0,    1,    2, ..., 5245, 5246, 5247])), TableProxy[A2626_1563320161_XXYY-Abell2626_corravgtim.ms](table, A2626_1563320161_XXYY-Abell2626_corravgtim.ms, ack=False,lockoptions=user,readonly=True,__executor_key__=A2626_1563320161_XXYY-Abell2626_corravgtim.ms), 'ANTENNA1', (<class 'tuple'>, []), dtype('int32'), array([[2.97976618e-05+0.j, 2.97976618e-05+0.j],
       [2.97971547e-05+0.j, 2.97971547e-05+0.j],
       [2.97871538e-05+0.j, 2.97871538e-05+0.j],
       [2.97830812e-05+0.j, 2.97830812e-05+0.j],
       [2.97830487e-05+0.j, 2.97830487e-05+0.j],
       [2.97717126e-05+0.j, 2.97717126e-05+0.j],
       [2.97716737e-05+0.j, 2.97716737e-05+0.j],
       [2.97699566e-05+0.j, 2.97699566e-05+0.j],
       [2.97674917e-05+
kwargs:    {}
Exception: MemoryError((52, 5248, 478, 2), dtype('complex128'))

```

**Comment by sjperkins** [comment]
With these changes, I ran the predict for about an hour with stable memory usage. The LD_PRELOAD jemalloc seems unnecessary.


```bash
$ LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so crystalball -sm ~/data/edb_2-sources-max.txt ~/data/testcb.ms/ -mf 0.1
Successful read/write open of default-locked table /home/sperkins/data/testcb.ms/: 26 columns, 349530 rows
-------------------------------------------
system RAM = 15.49 GB
nr of logical CPUs = 8
nr sources = 41427
nr rows    = 349530
nr chans   = 98
nr corrs   = 2
sources per chunk = 25 (auto setting)
rows per chunk    = 2575 (auto setting)
expected memory usage = 1.50 GB
-------------------------------------------
Nr sources        = 41427
-------------------------------------------
stokes.shape      = (41427, 4)
frequency.shape   = (98,)
/home/sperkins/work/ska/code/codex-africanus/africanus/rime/dask.py:41: PerformanceWarning: Increasing number of chunks by factor of 136
  dtype=infer_complex_dtype(lm, uvw, frequency))
brightness.shape  = (41427, 2)
phase.shape       = (41427, 349530, 98)
-------------------------------------------
Attempting phase-brightness einsum with "srf, si -> srfi"
jones.shape       = (41427, 349530, 98, 2)
-------------------------------------------
All sources are Delta functions
-------------------------------------------
distributed.comm.inproc - WARNING - Closing dangling queue in <InProc  local=inproc://10.8.10.20/5342/31 remote=inproc://10.8.10.20/5342/1>
distributed.comm.inproc - WARNING - Closing dangling queue in <InProc  local=inproc://10.8.10.20/5342/32 remote=inproc://10.8.10.20/5342/1>
distributed.comm.inproc - WARNING - Closing dangling queue in <InProc  local=inproc://10.8.10.20/5342/33 remote=inproc://10.8.10.20/5342/1>
distributed.comm.inproc - WARNING - Closing dangling queue in <InProc  local=inproc://10.8.10.20/5342/34 remote=inproc://10.8.10.20/5342/1>
distributed.comm.inproc - WARNING - Closing dangling queue in <InProc  local=inproc://10.8.10.20/5342/35 remote=inproc://10.8.10.20/5342/1>
[#                                       ] | 3% Completed |  4min 16.9sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.81 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 10min 47.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.16 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 10min 55.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.16 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.19 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 10min 56.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.16 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.20 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 10min 57.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.44 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.44 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 10min 57.8sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.26 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.26 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min  4.9sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.27 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.27 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min  6.1sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.16 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.00 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min  6.2sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.88 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.88 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min  8.3sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.51 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min 15.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.02 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.02 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min 15.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.91 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min 15.9sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.92 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min 16.9sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.54 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min 24.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.66 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.69 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min 32.4sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.58 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.60 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.29 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 11min 45.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 4.86 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 5% Completed | 12min  9.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.55 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 13min 46.5sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 4.91 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 13min 48.8sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.33 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 13min 50.0sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.52 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 13min 53.2sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.33 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 14min 27.3sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.03 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 16min 20.0sdistributed.worker - WARNING - gc.collect() took 1.731s. This is usually a sign that some tasks handle too many Python objects at the same time. Rechunking the work into smaller tasks might help.
distributed.worker - WARNING - Worker is at 80% memory usage. Pausing worker.  Process memory: 6.44 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 16min 22.0sdistributed.worker - WARNING - Worker is at 56% memory usage. Resuming worker. Process memory: 4.52 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 17min 14.1sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.03 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 18min 19.1sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 6.10 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.69 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 6% Completed | 18min 36.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.41 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 18min 54.0sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 4.85 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 19min 44.8sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.74 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.75 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 19min 45.9sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.54 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 19min 47.3sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.44 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 20min 25.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.32 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 20min 31.3sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.32 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 20min 52.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.15 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 20min 53.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 4.97 GB -- Worker memory limit: 8.00 GB
[##                                      ] | 7% Completed | 20min 59.8sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.07 GB -- Worker memory limit: 8.00 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.09 GB -- Worker memory limit: 8.00 GB
[###                                     ] | 8% Completed | 23min 25.9sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.23 GB -- Worker memory limit: 8.00 GB
[###                                     ] | 8% Completed | 23min 50.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.63 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  2min 32.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.52 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  2min 34.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.69 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  2min 35.6sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 4.89 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  2min 36.4sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.17 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  2min 36.5sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.17 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  2min 38.4sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.62 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  3min 22.7sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.68 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  3min 39.2sdistributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 5.32 GB -- Worker memory limit: 8.00 GB
[#######                                 ] | 17% Completed |  1hr  3min 53.3s
```

**Comment by sjperkins** [comment]
Right this is looking like a general garbage collection problem.

Putting in a worker `memory_limit=8GB` when creating the LocalCluster institutes a policy whereby the distributed scheduler aggressively garbage collects when the memory usage exceeds a certain limit. This is controlled by a `~/.config/dask/distributed.yaml` file (see the attached [zip](https://github.com/paoloserra/crystalball/files/3897469/distributed.zip))

The following options are present:

```yaml
    # Fractions of worker memory at which we take action to avoid memory blowup
    # Set any of the lower three values to False to turn off the behavior entirely
    memory:
      target: False  # target fraction to stay below
      spill: False  # fraction at which we spill to disk
      pause: 0.80  # fraction at which we pause worker threads
      terminate: 0.95  # fraction at which we terminate the worker
```

It's all a bit confusing because for the LocalCluster the limit seems to the limit for all workers (which are threads).

Regardless, instituting a memory limit means that if the memory used by the workers gets too much, dask distributed will enforce a garbage collection cycle. This temporarily hurts performance but means that the predict keeps going, even with > 40,000 sources.

**Comment by paoloserra** [comment]
I'm giving it a crack without the LD_PRELOAD . Getting flooded with messages like

`distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 13.65 GB -- Worker memory limit: 8.00 GB`

Is there a way to turn them off?

**Comment by sjperkins** [comment]
> I'm giving it a crack without the LD_PRELOAD . Getting flooded with messages like
> 
> `distributed.worker - WARNING - Memory use is high but worker has no data to store to disk. Perhaps some other process is leaking memory? Process memory: 13.65 GB -- Worker memory limit: 8.00 GB`
> 
> Is there a way to turn them off?

Look at the logging levels in distributed.yaml. Something like:

```yaml
  logging:
    distributed: error
    distributed.worker: error
```

**Comment by paoloserra** [comment]
> Look at the logging levels in distributed.yaml.

where's that supposed to live?

**Comment by sjperkins** [comment]
> where's that supposed to live?

I mentioned this in  https://github.com/paoloserra/crystalball/pull/33#issuecomment-559133527

> Putting in a worker memory_limit=8GB when creating the LocalCluster institutes a policy whereby the distributed scheduler aggressively garbage collects when the memory usage exceeds a certain limit. This is controlled by a ~/.config/dask/distributed.yaml file (see the attached zip)

File: Crystalball/crystalball.py [hunk 1]
```diff
@@ -160,13 +112,24 @@ def predict():
...
+    #                            n_workers=args.num_workers,
+    #                            threads_per_worker=1,
+    #                            memory_limit=None)
+    #     # Create distributed client
+    #     client = Client(cluster)
+
+    kwargs = dict(processes=False,
+                  n_workers=args.num_workers,
+                  threads_per_worker=1,
+                  memory_limit='8gb')
```
Comment (sjperkins) [inline Crystalball/crystalball.py]: @paoloserra Sorry the memory_limit is specified here.

Comment (paoloserra) [reply Crystalball/crystalball.py]: that's cool, I had seen that (and am not changing it for now)

**Comment by paoloserra** [comment]
> > where's that supposed to live?
> 
> I mentioned this in [#33 (comment)](https://github.com/paoloserra/crystalball/pull/33#issuecomment-559133527)

sorry, I had missed that!

File: Crystalball/crystalball.py (hunk 1 above)
Comment (sjperkins) [reply Crystalball/crystalball.py]: This should probably be `memory_fraction*total_memory`, *in this case*.

This is quite configurable and I think the definition of memory_limit changes when each worker is a process (in the above this is not the case `processes=False`).

In any case, I'll be intrigued to see if this runs through.

**Comment by paoloserra** [comment]
Not a huge deal, but I continue getting tons of those messages. This are the first few lines of my ~/.config/dask/distributed.yml

(note that it was all commented out when I first opened it)

```
distributed:
  version: 2
  logging:
    distributed: error
    distributed.client: error
    distributed.worker: error
    bokeh: critical
    # http://stackoverflow.com/questions/21234772/python-tornado-disable-logging-to-stderr
    tornado: critical
    tornado.application: error

#   scheduler:
#     allowed-failures: 3     # number of retries before a task is considered bad
#     bandwidth: 100000000    # 100 MB/s estimated worker-worker bandwidth
#     blocked-handlers: []
```

**Comment by sjperkins** [comment]
@paoloserra Set memory_limit to 50-75% of total RAM. 8GB was appropriate for my laptop. If they're very frequent the limit is too low.

Other that that, the yaml is based on the official docs. Give it a Google? I'm on my phone at the moment.

**Comment by paoloserra** [comment]
Bumped it up to 40GB but still plenty of warnings. Doesn't matter though, it's gonna run overnight anyway.

The **very good news** is that 30 min into the same large test above there's no signs of swap. Qualitatively, I see the CPUs being used a little less efficiently (as you expected), but who cares? :)

Will let you know when it finishes.

File: Crystalball/crystalball.py (hunk 1 above)
Comment (sjperkins) [reply Crystalball/crystalball.py]: There are a couple of options that can be set here. I'm experimenting with the following:

```python
    system_mem = float(psutil.virtual_memory()[0])
    memory_limit = int(args.memory_fraction*system_mem*2)
    memory_limit = int(0.5*system_mem)
    print("System Memory", system_mem, "Memory Limit", memory_limit)

    kwargs = dict(processes=False,
                  n_workers=args.num_workers,
                  threads_per_worker=1,
                  # Target fraction to stay below
                  memory_target_fraction=False,
                  # Don't spill to disk
                  memory_spill_fraction=False,
                  # Pause workers at 70% of memory_limit
                  memory_pause_fraction=0.7,
                  memory_limit=memory_limit
```

but can't get it to behave stably at the moment.

**Comment by sjperkins** [comment]
> Bumped it up to 40GB but still plenty of warnings. Doesn't matter though, it's gonna run overnight anyway.
> 
> The **very good news** is that 30 min into the same large test above there's no signs of swap. Qualitatively, I see the CPUs being used a little less efficiently (as you expected), but who cares? :)
> 
> Will let you know when it finishes.

Did it run through?

**Comment by paoloserra** [comment]
Hi @sjperkins ,

Yes, it kept on going. No swapping whatsoever throughout the run. I've killed it just now after 15 h of processing and ~50% progress (just because I need my machine back).

How do you want to proceed? Is this just a temporary patch or do you think this is a long-term solution?

**Comment by sjperkins** [comment]
> Hi @sjperkins ,
> 
> Yes, it kept on going. No swapping whatsoever throughout the run. I've killed it just now after 15 h of processing and ~50% progress (just because I need my machine back).
> 
> How do you want to proceed? Is this just a temporary patch or do you think this is a long-term solution?
> 
> **EDIT**: OK, just seen [#33 (comment)](https://github.com/paoloserra/crystalball/pull/33#discussion_r351634793) . Let me know if there's anything I can do.

OK glad to hear that, I do think its important to tune the worker memory profiles. I think we're running into garbage collector tuning, which can be tricky (see [here](https://www.cubrid.org/blog/how-to-tune-java-garbage-collection) for e.g.).

The first recommended step in garbage collection tuning is to reduce the number of allocated objects. I'm discussing this with people on my side. There may be a tradeoff in ease of RIME term composition vs memory + compute speed here that I need to think about.

On the other hand some reasonable worker memory limits as a stopgap would be great.

**Comment by paoloserra** [comment]
alright, thanks for the effort!

**Comment by sjperkins** [comment]
Closing in favour of https://github.com/paoloserra/crystalball/pull/34

**Pull request closed without merging**